### DISCUSSION TITLE
Can I use `sever.inject` in production ?

### DISCUSSION THREAD

**Discussion by Shudrum** [fastify/fastify#5131] [Q&A] [answered]
Hello!

I don't know if my question is at the right place or if I should ask the question on [fastify/light-my-request](https://github.com/fastify/light-my-request). Let me explain.

To be clear, I work on a very edge case project.

It's been years I use your project (Thank you ❤️ ), and now I need to make something quite complex.

To "simplify" my idea, let me explain a prototype:

I want an application, horizontally scalable, to start a "simple" server, listening for all methods, wildcard route.
I have in parallel a database with a list of hosts / routes / behavior.

When my main app's server receive a request for a specific `headers.host`, it checks if a server is loaded in memory for that host:
- **If not:** the application loads the routes from the database, create a server for that host and register the routes.
- **Then:** I would like the host's specific server to handle the request so the main server can answer.

This is why I would like to create a server per host, during the request (if it does not exists), and then process the request with `server.inject`, take the result, process it on the main server side with some pure business rules, and respond to the original request.

I do that because I don't know in advance the amount of hosts, and I don't want to expose a tons of port / have an unknown amount of instances running.

Is it such a bad idea ? or any advice ? I know I may loose some RPM, slowing down the process of the request if the server is not created on the instance receiving the request, but ... nothing more ...

Thank you!

**Comment by jsumners** [comment, marked as answer]
What I can say is that I have used `server.inject` to reduce repetition when I had handlers that needed to perform business logic, e.g. a `POST` endpoint using `server.inject` to get data from at `GET` endpoint before continuing.

**Reply by Shudrum** [reply]
Thank you for your feedback! Just knowing that someone has used it in production is reassuring.

**Reply by metcoder95** [reply]
Another thing to take a look at is on the `instance` if it does not call `fastify#close` you might be having memory leaking problems due to a small issue on `node:http`. Just closing it as soon as it is discarded should do the work

**Comment by Uzlopak** [comment]
I can also chime in. Imho. There are afaik some plugins using it actively and rely on it.

**Comment by Eomm** [comment]
It works BUT you must be aware of the downside IMHO:

- the body is serialized and parsed twice
- all the hooks will kick in, so an auth check will run twice as well

When I need the same output of a GET handlers I do this:

```js
app.get('/', getHandler)

app.post('/', async function postHandler(req, reply){
  const getResponse = await getHandler(req, reply)
  // do stuff
  return { done: true }
})

async function getHandler(req, reply) {
  // do stuff
  return { hello: 42 }
}
```

**Reply by melroy89** [reply]
I believe this is the only correct answer here.

**Comment by mcollina** [comment]
Yes you can. You might want to check out [Platformatic](https://docs.platformatic.dev) that with the combination of inject+undici+loaders allows for multiple Fastify apps to exists inside the same Node.js thread.
